server/auth/types/basic/basic_auth.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { AuthenticationHandler, SessionStorageFactory, IRouter, IClusterClient } from "../../../../../../src/core/server";
+import { SecurityPluginConfigType } from "../../..";
+import { SecuritySessionCookie } from "../../../session/security_cookie";
+import { CoreSetup } from "../../../../../../src/core/server";
+import _ from 'lodash';
+import { SecurityClient } from "../../../backend/opendistro_security_client";
+import { BasicAuthRoutes } from "./routes";
+
+export class AuthConfig {
+  constructor(
+    public readonly authType: string,
+    public readonly authHeaderName: string,
+    public readonly allowedAdditionalAuthHeaders: string[],
+    public readonly authenticateFunction: () => void,
+    public readonly validateAvailableTenants: boolean,
+    public readonly validateAvailableRoles: boolean) {
+  }
+}
+
+export class BasicAuthentication {
+  private static readonly AUTH_HEADER_NAME: string = 'authorization';
+  private static readonly ALLOWED_ADDITIONAL_AUTH_HEADERS: string[] = ['security_impersonate_as'];
+  private static readonly ROUTES_TO_IGNORE: string[] = [
+    '/bundles/app/security-login/bootstrap.js',
+    '/bundles/app/security-customerror/bootstrap.js',
+    '/',
+    '/app/login',
+    '/app/opendistro_login',
```

Comment:
Do we need 2 login endpoints?

Reply:
those are for my local testing, will remove them